manifold/src/csg_tree.cpp

```diff
@@ -0,0 +1,447 @@
+}
+
+std::shared_ptr<CsgNode> CsgOpNode::Transform(const glm::mat4x3 &m) const {
+  auto node = std::make_shared<CsgOpNode>();
+  node->children_ = children_;
+  node->op_ = op_;
+  node->transform_ = m * glm::mat4(transform_);
+  node->simplified = simplified;
+  node->flattened = flattened;
+  return node;
+}
+
+std::shared_ptr<CsgLeafNode> CsgOpNode::ToLeafNode() const {
+  if (cache_ != nullptr) return cache_;
+  if (children_.empty()) return nullptr;
+  // turn the children into leaf nodes
+  GetChildren();
+  Manifold::OpType op;
+  switch (op_) {
+    case CsgNodeType::UNION:
+      op = Manifold::OpType::ADD;
+      break;
+    case CsgNodeType::DIFFERENCE:
+      op = Manifold::OpType::SUBTRACT;
+      break;
+    case CsgNodeType::INTERSECTION:
+      op = Manifold::OpType::INTERSECT;
+      break;
+    default:
+      throw std::runtime_error("unreachable CSG operation");
+      break;
+  }
+  auto a = std::static_pointer_cast<CsgLeafNode>(children_[0])->GetImpl();
+  for (int i = 1; i < children_.size(); i++) {
+    auto b = std::static_pointer_cast<CsgLeafNode>(children_[i])->GetImpl();
+    Boolean3 boolean(*a, *b, op);
+    a = std::make_shared<Manifold::Impl>(boolean.Result(op));
+  }
+  children_.clear();
+  children_.push_back(std::make_shared<CsgLeafNode>(a));
+  // children_ must contain only one CsgLeafNode now, and its Transform will
+  // give CsgLeafNode as well
+  cache_ = std::dynamic_pointer_cast<CsgLeafNode>(
+      children_.front()->Transform(transform_));
+  return cache_;
+}
+
+/**
+ * Efficient boolean operation on a set of nodes utilizing commutativity of the
+ * operation. Only supports union and intersection.
+ */
+void CsgOpNode::BatchBoolean(
+    Manifold::OpType operation,
+    std::vector<std::shared_ptr<const Manifold::Impl>> &results) {
+  assert(operation != Manifold::OpType::SUBTRACT);
+  auto cmpFn = [](std::shared_ptr<const Manifold::Impl> a,
+                  std::shared_ptr<const Manifold::Impl> b) {
+    // invert the order because we want a min heap
+    return a->NumVert() > b->NumVert();
+  };
+
+  // apply boolean operations starting from smaller meshes
+  // the assumption is that boolean operations on smaller meshes is faster,
+  // due to less data being copied and processed
+  std::make_heap(results.begin(), results.end(), cmpFn);
+  while (results.size() > 1) {
+    std::pop_heap(results.begin(), results.end(), cmpFn);
+    auto a = std::move(results.back());
+    results.pop_back();
+    std::pop_heap(results.begin(), results.end(), cmpFn);
+    auto b = std::move(results.back());
+    results.pop_back();
+    // boolean operation
+    Boolean3 boolean(*a, *b, operation);
+    results.push_back(
+        std::make_shared<const Manifold::Impl>(boolean.Result(operation)));
+    std::push_heap(results.begin(), results.end(), cmpFn);
+  }
+}
+
+/**
+ * Efficient union operation on a set of nodes by doing Compose as much as
+ * possible.
+ */
+void CsgOpNode::BatchUnion() const {
```

Comment:
Okay, so these functions are still here, but just currently unused? Sounds good for now.